en/dev_log/ulog_file_format.md

```diff
@@ -41,233 +50,328 @@ The file consists of three sections:
-  ```c
-  struct message_add_logged_s {
-  	struct message_header_s header;
-  	uint8_t multi_id;
-  	uint16_t msg_id;
-  	char message_name[header.msg_size-3];
-  };
-  ```
-  `multi_id`: the same message format can have multiple instances, for example if the system has two sensors of the same type.
-  The default and first instance must be 0.
-  `msg_id`: unique id to match `message_data_s` data.
-  The first use must set this to 0, then increase it.
-  The same `msg_id` must not be used twice for different subscriptions, not even after unsubscribing.
-  `message_name`: message name to subscribe to.
-  Must match one of the `message_format_s` definitions.
-
-- 'R': unsubscribe a message, to mark that it will not be logged anymore (not used currently).
-  ```c
-  struct message_remove_logged_s {
-  	struct message_header_s header;
-  	uint16_t msg_id;
-  };
-  ```
+The message types in the _Data_ section are:
 
-- 'D': contains logged data.
-  ```
-  struct message_data_s {
-  	struct message_header_s header;
-  	uint16_t msg_id;
-  	uint8_t data[header.msg_size-2];
-  };
-  ```
-  `msg_id`: as defined by a `message_add_logged_s` message.
-  `data` contains the logged binary message as defined by `message_format_s`.
-  See above for special treatment of padding fields.
+1. [Subscription](#a-subscription-message)
+2. [Unsubscription](#r-unsubscription-message)
+3. [Logged Data](#d-logged-data-message)
+4. [Logged String](#l-logged-string-message)
+5. [Tagged Logged String](#c-tagged-logged-string-message)
+6. [Synchronization](#s-synchronization-message)
+7. [Dropout Mark](#o-dropout-message)
+8. [Information](#i-information-message)
+9. [Multi Information](#m-multi-information-message)
+10. [Parameter](#p-parameter-message)
+11. [Default Parameter](#q-default-parameter-message)
 
-- 'L': Logged string message, i.e. printf output.
-  ```
-  struct message_logging_s {
-  	struct message_header_s header;
-  	uint8_t log_level;
-  	uint64_t timestamp;
-  	char message[header.msg_size-9]
-  };
-  ```
-  `timestamp`: in microseconds, `log_level`: same as in the Linux kernel:
+#### `A`: Subscription Message
+
+Subscribe a message by name and give it an id that is used in [Logged data Message](#d-logged-data-message).
+This must come before the first corresponding [Logged data Message](#d-logged-data-message).
+
+```c
+struct message_add_logged_s {
+  struct message_header_s header; // msg_type = 'A'
+  uint8_t multi_id;
+  uint16_t msg_id;
+  char message_name[header.msg_size-3];
+};
+```
+
+- `multi_id`: the same message format can have multiple instances, for example if the system has two sensors of the same type. The default and first instance must be 0.
+- `msg_id`: unique id to match [Logged data Message](#d-logged-data-message) data. The first use must set this to 0, then increase it.
+  - The same `msg_id` must not be used twice for different subscriptions, not even after unsubscribing.
```

Comment:
I don't see why it should not be possible to re-use after unsubscribing.

Reply:
Fixed!